Quantum state classes

chriseclectic opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary

Adds quantum state classes to `quantum_info` package. These can be evolved by operator class objects, instructions or circuits.

### Details and comments

* Adds `QuantumState` abstract base class
* Adds `Statevector` class for complex statevector. Can be evolved by Operators and circuits containing gate instructions.
* Adds `DensityMatrix` class for density matrices. Can be evolved by Operators, QuantumChannels, and circuits containing gates, reset, and kraus instructions.

### TODO

* [x] Tests for Statevector
* [x] Tests for DensityMatrix
* ~~Integration with partial_trace and state_fidelity functions~~ (*follow up PR*)

chriseclectic commented:
@ajavadia ready for review

ajavadia commented on test/python/quantum_info/operators/channel/test_chi.py:
```diff
@@ -77,26 +77,26 @@ def test_evolve(self):
-        self.assertAllClose(chan._evolve(np.array(input_psi)), target_rho)
-        self.assertAllClose(chan._evolve(input_rho), target_rho)
-        self.assertAllClose(chan._evolve(np.array(input_rho)), target_rho)
+        self.assertAllClose(chan._evolve(input_psi).data, target_rho)
```
can you avoid using private methods in the tests? There are quite a lot of them. It should be possible to write tests using the public API.

chriseclectic replied:
These could be changed to `DensityMatrix(input_rho).evolve(chan)`

chriseclectic commented:
@ajavadia `_evolve` is removed from Operators and base class but it is a helper internal method in `QuantumChannels` that states classes use in their implementation of `evolve` for those objects. The `_evovle` method of channels isn't intended to be called by the user directly (thus the underscore).

ewinston commented on test/python/quantum_info/operators/channel/test_evolve.py:
```diff
@@ -41,10 +42,9 @@ def _unitary_to_other(self, rep, qubits_test_cases, repetitions):
             for _ in range(repetitions):
                 rho = self.rand_rho(dim)
```
There was some discussion about using random in tests. Did we come to an agreement about that?

chriseclectic replied:
I'm not aware of the outcome of that discussion, but I can say I found the random tests quite helpful for finding bugs

kdk replied:
More or less, the outcome was that tests that depend on randomness should set a seed, use a `RandomState`, and log the seed (see #1863). If you have tests that are intended to run continuously to look for bugs, they should live in `test/randomized`.

chriseclectic replied:
Added logging as per example in 1863

chriseclectic commented:
Updates

* Added `Statevector.from_label` and `DensityMatrix.from_label` class methods that construct from a label of Pauli eigenstates `0,1,+,-,r,l`.

* Removed `__init__` from instructions or circuits from class init and added `.from_instruction` class method that can initialize a Statevector or Density matrix from an instruction/circuit assuming the initial state is the all zero state (different initial states and be achieved the *standard* way of `Statevector(init_state).evolve(circuit)`).

chriseclectic commented:
@ajavadia @ewinston made the changes, should be good to go now
